engine/lib/worker/exit_controller.go

```diff
@@ -71,6 +127,14 @@ func (c *ExitController) PollExit() error {
 		return nil
 	}
 
+	c.callPrepareOnce.Do(func() {
```

Comment:
not familiar with framework, will the first poll block following call to poll? otherwise the first PollExit will enter callPrepareOnce but it may stuck at prepareExitFunc, and seconds PollExit will not wait prepareExitFunc so Exit finishes before CloseImpl

Reply:
Plus, `sync.Once` prevents the scenario you describe.